### DISCUSSION TITLE
Clarification on Selection Expansion in Output Mode ('o') for `select` Command

### DISCUSSION THREAD

**Discussion by mmxsrup** [YosysHQ/yosys#4674] [Q&A]
I’m looking for clarification on the behavior of the `select` command in Yosys, specifically when using output mode ('o'). I noticed that in the implementation of the `select_op_expand function`, the selection seems to expand from right to left, which appears counterintuitive for 'o' mode, where I expected the selection to expand from left (input) to right (output).

Here is the relevant code block in question:
https://github.com/YosysHQ/yosys/blob/38dae1d7e000e0fe587980ec211b827878472a0c/passes/cmds/select.cc#L458-L459

Shouldn’t this logic instead expand from left to right in 'o' mode, like this?
```cpp
if (mode != 'o' && selected_wires.count(conn_lhs[i].wire) && selected_members.count(conn_rhs[i].wire->name) == 0)
    lhs.selected_members[mod->name].insert(conn_rhs[i].wire->name), sel_objects++, max_objects--; 
```

Could someone confirm if this is the intended behavior in 'o' mode, or if the code should indeed expand from left to right in this case? Any clarification or corrections would be greatly appreciated.

**Comment by KrystalDelusion** [comment]
Not an answer, but possibly related; I have a pending `todo` about whether rule ordering in selections is a bug or a feature. Using [`code_examples/selections`](https://github.com/YosysHQ/yosys/tree/main/docs/source/code_examples/selections), calling `show y %ci*:-[CLK,S]:+$dff,$mux` gives 
![image](https://github.com/user-attachments/assets/26b7ff9e-f405-4b8b-af86-47c2b09711ae)
while `show y %ci*:+$dff,$mux:-[CLK,S]` gives 
![image](https://github.com/user-attachments/assets/615ba675-b85d-4c86-a4f0-b4b0091ae22d).

**Comment by povik** [comment]
> I expected the selection to expand from left (input) to right (output).

For connections LHS is the target net and RHS is the source net. In that light the expansion code reads correct to me.
